### PR TITLE
Build and publish multi-arch docker images on tag

### LINKED ISSUE #270: Update to go 1.16.x

### LINKED ISSUE #318: Build and publish docker images for dual arm64/amd64 support
https://hub.docker.com/r/honeycombio/refinery/tags?page=1&ordering=last_updated

### PR DESCRIPTION
**Pull request by MikeGoldsmith**
## Which problem is this PR solving?
When publishing to docker, our images should be built and be runnable on multiple architectures. This PR updates our docker image build process to use [Google KO](https://github.com/google/ko) to build the image.
- Resolves #318 
- Resolves #270 

## Short description of the changes
- update go.mod to use v.1.16
- replace docker orb with new `publish_docker` job
- use ko to build multi-arch docker image
- add build/build_docker.sh to manage executing ko command

### CODE REVIEW

File: build/build_docker.sh [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+PLATFORM="${PLATFORM:-linux/amd64,linux/arm64,darwin/amd64}"
+
+unset GOOS
+unset GOARCH
+export KO_DOCKER_REPO=${KO_DOCKER_REPO:-ko.local}
+export GOFLAGS="-ldflags=-X=main.version=${VERSION}"
+export SOURCE_DATE_EPOCH=$(date +%s)
+# shellcheck disable=SC2086
+ko publish \
+  --tags "head,${VERSION}" \
```
Comment (vreynolds) [inline build/build_docker.sh:14]: should this be `latest` tag?

Comment (MikeGoldsmith) [reply build/build_docker.sh]: Yep, we should be tagging latest and whatever CIRCLECI tag we see 👍🏻

**Review by vreynolds — APPROVED** [review]

File: .circleci/config.yml [hunk 2]
```diff
@@ -124,6 +123,18 @@ jobs:
             if [[ -z "$version" ]] ; then version=${CIRCLE_SHA1:0:7}; fi
             aws s3 sync ~/artifacts s3://honeycomb-builds/honeycombio/refinery/$version/
 
+  publish_docker:
+    executor: linuxgo
+    steps:
+      - checkout
+      - setup_remote_docker
+      - run: go install github.com/google/ko@latest
```
Comment (robbkidd) [inline .circleci/config.yml:131]: Recommendation: cache the ko install.

File: .circleci/config.yml [hunk 3]
```diff
@@ -124,6 +123,18 @@ jobs:
...
+      - run:
+          name: "Publish multi-arch docker image"
+          command: |
+            echo "${DOCKER_PASSWORD}" | docker login -u "${DOCKER_USERNAME}" --password-stdin;
+            VERSION=${CIRCLE_TAG} KO_DOCKER_REPO=honeycombio build/build_docker.sh
```
Comment (robbkidd) [inline .circleci/config.yml:136]: Recommendation: also perform a local ko build during the test or build job so we can see build errors earlier (on PRs and not just on taggings).

File: build/build_docker.sh [hunk 4]
```diff
@@ -0,0 +1,20 @@
+set -o nounset
+set -o pipefail
```
Comment (robbkidd) [inline build/build_docker.sh:2]: Recommendation: `set -x` for the script to echo each command during execution to facilitate debugging. I don't (yet?) see sensitive values in the commands or output.

```suggestion
set -o pipefail
set -o xtrace
```

**Review by robbkidd — COMMENTED** [review]
Carrying over some @ismith's recommendations in [similar changes to honeymarker](https://github.com/honeycombio/honeymarker/pull/27#pullrequestreview-766975917)

File: .circleci/config.yml [hunk 5]
```diff
@@ -124,6 +123,25 @@ jobs:
             if [[ -z "$version" ]] ; then version=${CIRCLE_SHA1:0:7}; fi
             aws s3 sync ~/artifacts s3://honeycomb-builds/honeycombio/refinery/$version/
 
+  publish_docker:
+    executor: linuxgo
+    steps:
+      - restore_cache:
+          keys:
+            googkleko
```
Comment (robbkidd) [inline .circleci/config.yml:131]: Is the extra "k" in the key name meaningful? The save and restore keys match, so it works, just wondering if I'm missing meaning or if this was a typo.

File: .circleci/config.yml [hunk 6]
```diff
@@ -124,6 +123,25 @@ jobs:
...
+      - run: go install github.com/google/ko@latest
+      - save_cache:
+          key: googkleko
```
Comment (robbkidd) [inline .circleci/config.yml:134]:
```suggestion
          key: googleko
```

**Pull request merged**